runtime/commonMain/src/kotlinx/serialization/TypeTokenResolving.kt

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+import kotlinx.serialization.internal.*
+import kotlin.reflect.*
+
+/**
+ * Reified version of [serializerByKType], provided for convenience.
+ * [KType] is obtained via [typeOf] function, consult its documentation
+ * to understand limitations.
+ *
+ * [typeOf] and [KType] API currently don't work on Kotlin/JS.
+ */
+@Suppress("UNCHECKED_CAST", "NO_REFLECTION_IN_CLASS_PATH")
+@ImplicitReflectionSerializer
+public inline fun <reified T> serializer(): KSerializer<T> {
+    return serializerByKType(typeOf<T>()) as KSerializer<T>
+}
+
+/**
+ * This method is able to construct serializer for a given [type].
+ * Since [KType], in contrary to [KClass], contains information about generic type parameters,
+ * it is possible to construct serializers for collections using this method.
+ * However, user-defined generic classes are not supported for now.
+ *
+ * Keep in mind that this is a 'heavy' call, so result probably should be cached somewhere else.
+ *
+ * [typeOf] and [KType] API currently don't work on Kotlin/JS.
+ *
+ * @see typeOf
+ */
+@Suppress("UNCHECKED_CAST", "NO_REFLECTION_IN_CLASS_PATH", "UNSUPPORTED")
+@ImplicitReflectionSerializer
+public fun serializerByKType(type: KType): KSerializer<Any?> {
+    fun serializerByKTypeImpl(type: KType): KSerializer<Any> {
+        val rootClass = when (val t = type.classifier) {
+            is KClass<*> -> t
+            else -> error("Only KClass supported as classifier, got $t")
+        } as KClass<Any>
+        return when {
+            type.arguments.isEmpty() -> rootClass.serializer()
+            else -> {
+                val args = type.arguments
+                    .map { requireNotNull(it.type) { "Star projections are not allowed" } }
+                    .map(::serializerByKType)
+                // Array is not supported, see KT-32839
+                when (rootClass) {
+                    List::class, ArrayList::class -> ArrayListSerializer(args[0])
```

Comment:
Do we want to support all potential list supertypes by inspecting superclasses of the given type?

Reply:
What do you mean by supertypes? Any or something?

If you mean subtypes, the design decision was especially no – we provide default list serializers only for default list types. Moreover, KType does not contain `isSubtype/isSupertype` API

Reply:
Yeah, I've meant subtypes.
Ok, works for me.

>Moreover, KType does not contain isSubtype/isSupertype API

`KClass.superclass` + casting to `KClass` should work